geoportal/c2cgeoportal_geoportal/views/login.py

```diff
@@ -0,0 +1,340 @@
+    def loginform403(self):
+        if self.request.authenticated_userid:
+            return HTTPUnauthorized()  # pragma: no cover
+
+        set_common_headers(self.request, "login", NO_CACHE)
+
+        return {"lang": self.lang, "came_from": self.request.path, "two_fa": self.two_factor_auth}
+
+    @view_config(route_name="loginform", renderer="login.html")
+    def loginform(self):
+        set_common_headers(self.request, "login", PUBLIC_CACHE)
+
+        return {
+            "lang": self.lang,
+            "came_from": self.request.params.get("came_from") or "/",
+            "two_fa": self.two_factor_auth,
+        }
+
+    def _validate_2fa_totp(self, user, otp: str) -> bool:
+        if pyotp.TOTP(user.tech_data.get("2fa_totp_secret", "")).verify(otp):
+            return True
+        return False
+
+    @view_config(route_name="login")
+    def login(self):
+        self._referer_log()
+
+        login = self.request.POST.get("login")
+        password = self.request.POST.get("password")
+        if login is None or password is None:  # pragma nocover
+            raise HTTPBadRequest("'login' and 'password' should be available in request params.")
+        username = self.request.registry.validate_user(self.request, login, password)
+        if username is not None:
+            user = models.DBSession.query(static.User).filter(static.User.username == username).one()
+            if self.two_factor_auth:
+                if "2fa_totp_secret" not in user.tech_data:
+                    user.is_password_changed = False
+                if not user.is_password_changed:
+                    user.tech_data["2fa_totp_secret"] = pyotp.random_base32()
+                    return set_common_headers(
+                        self.request,
+                        "login",
+                        NO_CACHE,
+                        response=Response(
+                            json.dumps(
+                                {
+                                    "username": user.username,
+                                    "is_password_changed": False,
+                                    "two_factor_enable": True,
+                                    "two_factor_totp_secret": user.tech_data["2fa_totp_secret"],
+                                    "otp_uri": pyotp.TOTP(user.tech_data["2fa_totp_secret"]).provisioning_uri(
+                                        user.email, issuer_name=self.two_factor_issuer_name
+                                    ),
+                                }
+                            ),
+                            headers=(("Content-Type", "text/json"),),
+                        ),
+                    )
+                otp = self.request.POST.get("otp")
+                if otp is None:
+                    raise HTTPBadRequest("The second factor is missing.")
+                if not self._validate_2fa_totp(user, otp):
+                    LOG.info("The second factor is wrong for user '%s'.", user)
```

Comment:
Risks of displaying the hashed password of the user, there... I don't know.
To be sure, you can put `user.username` instead...

Reply:
Yes, in another pull request :-)